### ISSUE TITLE
MarketMap fails to update

### ISSUE THREAD

**Issue by acelere** [bqplot/bqplot#1278]
**Describe the bug**
Update MarketMap.map.names does not work. 

**To Reproduce**
I am running Jupyterhub using dockerspawner.
In my "old" container, it worked perfectly. 
After rebuilding the container, the bug appeared.
I tried to roll back versions as much as I could to pinpoint the culprit, but was not able to find it.
This is the pip freeze of both containers. Column on the left works, on the right does not:

Jupyter Prod - working |   | Jupyter Dev - NOT working
-- | -- | --
$ pip freeze |   | $ pip freeze
absl-py==0.11.0 |   |  
aiohttp==3.7.3 |   | aiohttp==3.7.3
alabaster==0.7.12 |   |  
alembic==1.4.3 | diff | alembic==1.5.4
anyio==2.0.2 | diff | anyio==2.1.0
argon2-cffi==20.1.0 |   | argon2-cffi==20.1.0
async-generator==1.10 |   | async-generator==1.10
async-timeout==3.0.1 |   | async-timeout==3.0.1
attrs==20.3.0 |   | attrs==20.3.0
Babel==2.9.0 |   |  
backcall==0.2.0 |   | backcall==0.2.0
beautifulsoup4==4.9.3 |   |  
bleach==3.2.1 | diff | bleach==3.3.0
bqplot==0.12.21 |   | bqplot==0.12.21
certifi==2020.12.5 |   | certifi==2020.12.5
certipy==0.1.3 |   | certipy==0.1.3
cffi==1.14.4 | diff | cffi==1.14.5
chardet==3.0.4 |   | chardet==3.0.4
click==7.1.2 |   | click==7.1.2
colorama==0.4.4 |   |  
cryptography==3.3.1 | diff | cryptography==3.4.5
cvxpy==1.1.7 |   |  
cycler==0.10.0 |   | cycler==0.10.0
debugpy==1.2.1 |   |  
decorator==4.4.2 |   | decorator==4.4.2
defusedxml==0.6.0 |   | defusedxml==0.6.0
docutils==0.16 |   |  
ecos==2.0.7.post1 |   |  
entrypoints==0.3 |   | entrypoints==0.3
flatbuffers==1.12 |   |  
future==0.18.2 |   |  
gitdb==4.0.5 |   |  
GitPython==3.1.12 |   |  
h5py==3.1.0 |   |  
idna==2.10 |   | idna==2.10
imagesize==1.2.0 |   |  
importlib-metadata==3.4.0 |   |  
ipydatawidgets==4.2.0 |   |  
ipyfilechooser==0.4.1 |   | ipyfilechooser==0.4.1
ipykernel==5.4.3 |   | ipykernel==5.4.3
ipython==7.19.0 | diff | ipython==7.20.0
ipython-genutils==0.2.0 |   | ipython-genutils==0.2.0
ipyvolume==0.5.2 |   |  
ipywebrtc==0.5.0 |   |  
ipywidgets==7.6.3 |   | ipywidgets==7.6.3
jax==0.2.8 |   |  
jaxlib==0.1.59 |   |  
jedi==0.17.0 | diff | jedi==0.18.0
jhsingle-native-proxy==0.6.1 |   | jhsingle-native-proxy==0.6.1
Jinja2==2.11.2 | diff | Jinja2==2.11.3
json5==0.9.5 |   | json5==0.9.5
jsonschema==3.2.0 |   | jsonschema==3.2.0
jupyter-book==0.9.1 |   |  
jupyter-cache==0.4.1 |   |  
jupyter-client==6.1.11 |   | jupyter-client==6.1.11
jupyter-core==4.7.0 | diff | jupyter-core==4.7.1
jupyter-server==1.2.2 | diff | jupyter-server==1.3.0
jupyter-sphinx==0.3.1 |   |  
jupyter-telemetry==0.1.0 |   | jupyter-telemetry==0.1.0
jupyterhub==1.3.0 |   | jupyterhub==1.3.0
jupyterlab==2.2.9 |   | jupyterlab==2.2.9
jupyterlab-pygments==0.1.2 |   | jupyterlab-pygments==0.1.2
jupyterlab-server==1.2.0 |   | jupyterlab-server==1.2.0
jupyterlab-widgets==1.0.0 |   | jupyterlab-widgets==1.0.0
jupytext==1.7.1 |   |  
kiwisolver==1.3.1 |   | kiwisolver==1.3.1
latexcodec==2.0.1 |   |  
Mako==1.1.4 |   | Mako==1.1.4
markdown-it-py==0.5.8 |   |  
MarkupSafe==1.1.1 |   | MarkupSafe==1.1.1
matplotlib==3.3.3 |   | matplotlib==3.3.3
metakernel==0.27.5 |   |  
mistune==0.8.4 |   | mistune==0.8.4
mpmath==1.1.0 |   |  
multidict==5.1.0 |   | multidict==5.1.0
myst-nb==0.10.2 |   |  
myst-parser==0.12.10 |   |  
nbclient==0.5.1 | diff | nbclient==0.5.2
nbconvert==6.0.7 |   | nbconvert==6.0.7
nbdime==2.1.0 |   |  
nbformat==5.1.2 |   | nbformat==5.1.2
nest-asyncio==1.4.3 | diff | nest-asyncio==1.5.1
nested-lookup==0.2.21 |   |  
notebook==6.2.0 |   | notebook==6.2.0
numpy==1.19.5 |   | numpy==1.19.5
oauthlib==3.1.0 |   | oauthlib==3.1.0
octave-kernel==0.32.0 |   |  
opt-einsum==3.3.0 |   |  
oset==0.1.3 |   |  
osqp==0.6.2.post0 |   |  
packaging==20.8 | diff | packaging==20.9
pamela==1.0.0 |   | pamela==1.0.0
pandas==1.2.0 |   | pandas==1.2.0
pandocfilters==1.4.3 |   | pandocfilters==1.4.3
parso==0.8.1 |   | parso==0.8.1
pexpect==4.8.0 |   | pexpect==4.8.0
pickleshare==0.7.5 |   | pickleshare==0.7.5
Pillow==8.1.0 |   | Pillow==8.1.0
prometheus-client==0.9.0 |   | prometheus-client==0.9.0
prompt-toolkit==3.0.10 | diff | prompt-toolkit==3.0.16
ptyprocess==0.7.0 |   | ptyprocess==0.7.0
pybtex==0.23.0 |   |  
pybtex-docutils==1.0.0 |   |  
pycparser==2.20 |   | pycparser==2.20
pydata-sphinx-theme==0.4.1 |   |  
Pygments==2.7.4 | diff | Pygments==2.8.0
pyOpenSSL==20.0.1 |   | pyOpenSSL==20.0.1
pyparsing==2.4.7 |   | pyparsing==2.4.7
pyrsistent==0.17.3 |   | pyrsistent==0.17.3
python-dateutil==2.8.1 |   | python-dateutil==2.8.1
python-editor==1.0.4 |   | python-editor==1.0.4
python-json-logger==2.0.1 |   | python-json-logger==2.0.1
pythreejs==2.2.1 |   |  
pytz==2020.5 | diff | pytz==2021.1
PyYAML==5.3.1 |   |  
pyzmq==21.0.1 |   | pyzmq==21.0.1
qdldl==0.1.5.post0 |   |  
requests==2.25.1 |   | requests==2.25.1
ruamel.yaml==0.16.12 |   | ruamel.yaml==0.16.12
scikit-aero==0.1 |   |  
scipy==1.6.0 |   |  
scs==2.1.2 |   |  
seaborn==0.11.1 |   |  
Send2Trash==1.5.0 |   | Send2Trash==1.5.0
simpervisor==0.4 |   | simpervisor==0.4
six==1.15.0 |   | six==1.15.0
smmap==3.0.4 |   |  
sniffio==1.2.0 |   | sniffio==1.2.0
snowballstemmer==2.0.0 |   |  
soupsieve==2.1 |   |  
Sphinx==3.4.3 |   |  
sphinx-book-theme==0.0.39 |   |  
sphinx-comments==0.0.3 |   |  
sphinx-copybutton==0.3.1 |   |  
sphinx-panels==0.5.2 |   |  
sphinx-thebe==0.0.8 |   |  
sphinx-togglebutton==0.2.3 |   |  
sphinxcontrib-applehelp==1.0.2 |   |  
sphinxcontrib-bibtex==1.0.0 |   |  
sphinxcontrib-devhelp==1.0.2 |   |  
sphinxcontrib-htmlhelp==1.0.3 |   |  
sphinxcontrib-jsmath==1.0.1 |   |  
sphinxcontrib-qthelp==1.0.3 |   |  
sphinxcontrib-serializinghtml==1.1.4 |   |  
SQLAlchemy==1.3.22 | diff | SQLAlchemy==1.3.23
sympy==1.7.1 |   |  
terminado==0.9.2 |   | terminado==0.9.2
testpath==0.4.4 |   | testpath==0.4.4
toml==0.10.2 |   |  
torch==1.7.1 |   |  
torchvision==0.8.2 |   |  
tornado==6.1 |   | tornado==6.1
traitlets==5.0.5 |   | traitlets==5.0.5
traittypes==0.2.1 |   | traittypes==0.2.1
typing-extensions==3.7.4.3 |   | typing-extensions==3.7.4.3
uncertainties==3.1.5 |   |  
urllib3==1.26.2 | diff | urllib3==1.26.3
voila==0.2.6 |   | voila==0.2.6
wcwidth==0.2.5 |   | wcwidth==0.2.5
webencodings==0.5.1 |   | webencodings==0.5.1
widgetsnbextension==3.5.1 |   | widgetsnbextension==3.5.1
xarray==0.16.2 |   |  
xeus-python==0.9.1 |   |  
xlrd==2.0.1 |   |  
yarl==1.6.3 |   | yarl==1.6.3
zipp==3.4.0 |   |  

This is the code to reporduce:

Example code:
```
#create MarketMap
import pandas as pd
from bqplot.market_map import MarketMap
from ipywidgets import Layout

class ParameterMap:
    def __init__(self, sub_sampled_data, title, map_groups):
        self.sub_sampled_data = sub_sampled_data
        self.map_names = list(self.sub_sampled_data.columns)
        self.map_codes = [i for i in range(len(self.map_names))]
        self.map_groups = []
        for key in map_groups.keys():
            self.map_groups.append(map_groups.get(key))
        
        self.map = MarketMap(names=self.map_names, groups=self.map_groups,     
                               layout=Layout(min_width='50px', min_height='70px'),
                                 enable_hover=False, cols=3,
                            map_margin={'top':50, 'bottom':0, 'left':0, 'right':25})

        self.map.font_style = {'font-size': '10px', 'fill':'white'}
        self.map.title = title
        self.map.title_style = {'fill': 'Red'}
        
    def update_map(self, sub_sampled_data, map_groups):
        self.sub_sampled_data = sub_sampled_data
        self.map_names = list(self.sub_sampled_data.columns)
        self.map_codes = [i for i in range(len(self.map_names))]
        self.map_groups = []
        for key in map_groups.keys():
            self.map_groups.append(map_groups.get(key))
        
        self.map.names = self.map_names
        self.map.groups = self.map_groups

data_label = 'test_dat0'
data_length = 10
test_data = {}
for i in range(data_length):
    test_data[f'{data_label}_{i}'] = [1,2,3,4] 
test_data = pd.DataFrame(test_data)
sub_sampled_data = test_data
map_groups = {}
for p in sub_sampled_data.columns:
    map_groups[p] = 'FTI'
slicemap = ParameterMap(sub_sampled_data, 'Parameter Map', map_groups)
slicemap.map
```
and on another notebook cell:

```
#update the MarketMap data
data_label = 'test_dat1'
data_length = 30
test_data = {}
for i in range(data_length):
    test_data[f'{data_label}_{i}'] = [1,2,3,4] 
test_data = pd.DataFrame(test_data)
sub_sampled_data = test_data
map_groups = {}
for p in sub_sampled_data.columns:
    map_groups[p] = 'FTI'
slicemap.update_map(sub_sampled_data, map_groups)
```
Upon looking at the console for errors, there is a JS error "...is not a function" pointing to MarketMap.js:426 (see screen shot below)

**Expected behavior**
MarketMap to update with new data (in the example code, update with 30 cells)...

**Screenshots**
Working configuration:
![market_map_ok](https://user-images.githubusercontent.com/12059585/108216253-41a1e600-7100-11eb-9f0b-16bae74f7b09.gif)

NOT WORKING configuration:
![market_map_not_ok](https://user-images.githubusercontent.com/12059585/108216300-51212f00-7100-11eb-88bd-5c49c3fbb72d.gif)

**Additional context**
tested on Chrome and Firefox with same results.
The issue can also be observed if I create a brand new container, with the latest versions for all packages, including Jupyterlab 3.0.7

**Comment by acelere** [comment]
I did some further investigation and there is something fundamentally wrong.
MarketMap does not work at all.

I built a minimalist container with the following Dockerfile (I run Jupyterhub):

```
FROM python:3.9.1

RUN apt-get update

RUN curl -sL https://deb.nodesource.com/setup_15.x | bash -
RUN apt-get install -y nodejs


RUN pip3 install --upgrade pip

RUN pip3 install \
    notebook \
    jupyterhub \
    jupyterlab \
    numpy \
    pandas \
    ipywidgets \
    traitlets \
    bqplot


RUN jupyter labextension install @jupyter-widgets/jupyterlab-manager



# create a user, since we don't want to run as root
RUN useradd -m jovyan
RUN usermod -a -G users jovyan
# Add user to staff to enable R packages install
RUN usermod -a -G staff jovyan
ENV HOME=/home/jovyan
WORKDIR $HOME

USER jovyan

RUN mkdir /home/jovyan/work
RUN mkdir /home/jovyan/work/data
RUN chown -R jovyan /home/jovyan/work


CMD ["jupyterhub-singleuser"]

```

This is what I get after logging in Jupyterhub:

```
$ jupyter labextension list
JupyterLab v3.0.8
/usr/local/share/jupyter/labextensions
        bqplot v0.5.22 enabled OK
        @jupyter-widgets/jupyterlab-manager v3.0.0 enabled OK (python, jupyterlab_widgets)

Other labextensions (built into JupyterLab)
   app dir: /usr/local/share/jupyter/lab

$ pip freeze
alembic==1.5.4
anyio==2.1.0
argon2-cffi==20.1.0
async-generator==1.10
attrs==20.3.0
Babel==2.9.0
backcall==0.2.0
bleach==3.3.0
bqplot==0.12.22
certifi==2020.12.5
certipy==0.1.3
cffi==1.14.5
chardet==4.0.0
cryptography==3.4.6
decorator==4.4.2
defusedxml==0.6.0
entrypoints==0.3
idna==2.10
ipykernel==5.4.3
ipython==7.20.0
ipython-genutils==0.2.0
ipywidgets==7.6.3
jedi==0.18.0
Jinja2==2.11.3
json5==0.9.5
jsonschema==3.2.0
jupyter-client==6.1.11
jupyter-core==4.7.1
jupyter-server==1.4.0
jupyter-telemetry==0.1.0
jupyterhub==1.3.0
jupyterlab==3.0.8
jupyterlab-pygments==0.1.2
jupyterlab-server==2.2.1
jupyterlab-widgets==1.0.0
Mako==1.1.4
MarkupSafe==1.1.1
mistune==0.8.4
nbclassic==0.2.6
nbclient==0.5.2
nbconvert==6.0.7
nbformat==5.1.2
nest-asyncio==1.5.1
notebook==6.2.0
numpy==1.20.1
oauthlib==3.1.0
packaging==20.9
pamela==1.0.0
pandas==1.2.2
pandocfilters==1.4.3
parso==0.8.1
pexpect==4.8.0
pickleshare==0.7.5
prometheus-client==0.9.0
prompt-toolkit==3.0.16
ptyprocess==0.7.0
pycparser==2.20
Pygments==2.8.0
pyOpenSSL==20.0.1
pyparsing==2.4.7
pyrsistent==0.17.3
python-dateutil==2.8.1
python-editor==1.0.4
python-json-logger==2.0.1
pytz==2021.1
pyzmq==22.0.3
requests==2.25.1
ruamel.yaml==0.16.12
Send2Trash==1.5.0
six==1.15.0
sniffio==1.2.0
SQLAlchemy==1.3.23
terminado==0.9.2
testpath==0.4.4
tornado==6.1
traitlets==5.0.5
traittypes==0.2.1
urllib3==1.26.3
wcwidth==0.2.5
webencodings==0.5.1
widgetsnbextension==3.5.1
```
and this is the behavior that I see when I try to run the example code:
![market_map_not_ok2](https://user-images.githubusercontent.com/12059585/108549248-dfd9ab80-72ba-11eb-9b23-e7ea9c8a0741.gif)

Notice the error codes on the console and also the fact that clicking on a MarketMap cell to select it does not work. 
The expected behavior would be something like this (from running an old container that I have saved):
![market_map_ok2](https://user-images.githubusercontent.com/12059585/108549302-f7b12f80-72ba-11eb-82d8-fbe615719164.gif)

I wonder if this has anything to do with issue #1245 ...
@maartenbreddels , would you have a look?

Let me know if I can help as well...

**Comment by acelere** [comment]
Fixed by release 0.12.23
